src/main/java/me/hydos/rosella/debug/VulkanDebugCallback.java

```diff
@@ -0,0 +1,93 @@
+package me.hydos.rosella.debug;
+
+import org.jetbrains.annotations.NotNull;
+import org.lwjgl.vulkan.VK10;
+import org.lwjgl.vulkan.VkDebugUtilsMessengerCallbackDataEXT;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class VulkanDebugCallback {
+
+    private final Set<Callback> callbacks = new ConcurrentSkipListSet<>();
+
+    public VulkanDebugCallback() {
+    }
+
+    public void registerCallback(Callback cb) {
+        this.callbacks.add(cb);
+    }
+
+    public void removeCallback(Callback cb) {
+        this.callbacks.remove(cb);
+    }
+
+    public void destroy() {
+        // The callback is registered during instance creation so we dont have to destroy anything
+    }
+
+    public int vulkanCallbackFunction(int severityBits, int typeBits, long pCallbackData, long pUserData) {
+        try {
+            VkDebugUtilsMessengerCallbackDataEXT callbackData = VkDebugUtilsMessengerCallbackDataEXT.create(pCallbackData);
+            MessageSeverity severity = MessageSeverity.fromBits(severityBits);
+            MessageType type = MessageType.fromBits(typeBits);
+
+            callbacks.forEach(cb -> cb.call(severity, type, callbackData));
+
+        } catch (Exception ex) {
+            // TODO log?
+        }
+        return VK10.VK_FALSE;
+    }
+
+    public static abstract class Callback implements Comparable<Callback> {
+        // Needed as a value to compare to for the skip list
+        private static final AtomicInteger _nextId = new AtomicInteger(1);
+        private final int _id;
+
+        protected final AtomicInteger severityMask;
```

Comment:
Why atomic? They're never updated

Reply:
Every time you create a new callback it hast to get a new unique id. Since allocations can happen concurrently the nextId must be an atomic.

Reply:
I meant `severityMask` not `_nextId`

Reply:
Oh that is because you should be able to change the mask at runtime